### PR TITLE
feat: add support for git refs for vendir

### PR DESCRIPTION
**Pull request by knechtionscoding**
## Changes

Adds support for gitRefs (but not revisionSelections/Ranges) to the Vendir Manager. Vendir supports lots of different upstream/sources this is a commonly used one.

## Context

After merging support for helmCharts via Vendir this PR adds support for gitRefs as well

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/vendir/__fixtures__/git-contents.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+apiVersion: vendir.k14s.io/v1alpha1
+kind: Config
+directories:
+- path: vendor
+  contents:
+  - path: custom-repo-custom-version
```
Comment (viceice) [inline lib/modules/manager/vendir/__fixtures__/git-contents.yaml:7]: please join fixtures, so we've less files and less reads. we usually have one big fixture with all supported valid values and a single test which verified it. so adding new supported values needs less test code changes.

File: lib/modules/manager/vendir/artifacts.spec.ts [hunk 2]
```diff
@@ -476,7 +475,7 @@ describe('modules/manager/vendir/artifacts', () => {
           newPackageFileContent: vendirFile,
           config: {
             ...config,
-            constraints: { vendir: '0.35.0', helm: '3.17.0' },
+            constraints: { vendir: '0.35.0', helm: '3.17.0', git: '2.33.0' },
```
Comment (viceice) [inline lib/modules/manager/vendir/artifacts.spec.ts:478]: remove git. we can't install a specific version on the fly.

File: lib/modules/manager/vendir/types.ts [hunk 3]
```diff
@@ -1,36 +0,0 @@
-import type { HostRule } from '../../../types';
```
Comment (viceice) [inline lib/modules/manager/vendir/types.ts:1]: was it already unused? move to separate PR

File: lib/modules/manager/vendir/schema.ts [hunk 4]
```diff
@@ -16,7 +22,8 @@ export const HelmChart = z.object({
 
 export const Contents = z.object({
   path: z.string(),
-  helmChart: HelmChart,
+  helmChart: HelmChart.optional(),
+  git: GitRef.optional(),
```
Comment (viceice) [inline lib/modules/manager/vendir/schema.ts:26]: are both allowed? if not we should create two different objects.

File: lib/modules/manager/vendir/types.ts (hunk 3 above)
Comment (knechtionscoding) [reply lib/modules/manager/vendir/types.ts:1]: Yes. It's removal was supposed to be contained in the previous vendir PR, but that change got missed.

Comment (knechtionscoding) [reply lib/modules/manager/vendir/types.ts]: Will move to a separate PR

File: lib/modules/manager/vendir/schema.ts (hunk 4 above)
Comment (viceice) [reply lib/modules/manager/vendir/schema.ts:26]: ok, You misunderstood. are helm chart and git are allowed at same time on same object?

```yaml
  - path: XYZ
    helmChart: ...
    git: ...
```

Comment (knechtionscoding) [reply lib/modules/manager/vendir/schema.ts:26]: Ah, no, two different objects. Okay, I'll update.

Comment (viceice) [reply lib/modules/manager/vendir/schema.ts:26]: https://github.com/renovatebot/renovate/blob/b33ce14e88756dcc7e50bfb032529c238e7d0c92/lib/modules/platform/github/issue.ts#L27

use a union type 🙃

**Review by viceice — COMMENTED** [review]
can you please make the fixtures condensed in a separate PR? so we've less changes for the new feature.

**Comment by knechtionscoding** [comment]
> can you please make the fixtures condensed in a separate PR? so we've less changes for the new feature.

You had asked me to condense the fixtures in a previous comment. So it feels like doing 1/3 of the condensing here just because and then a second PR is... Odd? Either I follow the pattern for vendir from the previous PR of having separate fixtures and will condense if desired completely in a separate PR. Or I'll do it now.

But leaving it part way done/inconsistent doesn't seem right.

File: lib/modules/manager/vendir/schema.ts (hunk 4 above)
Comment (knechtionscoding) [reply lib/modules/manager/vendir/schema.ts]: Updated to use a union type

File: lib/modules/manager/vendir/readme.md [hunk 5]
```diff
@@ -1,8 +1,10 @@
-Renovate supports updating Helm Chart references in vendir.yml via the [vendir](https://carvel.dev/vendir/) tool. Renovate requires the presence of a [vendir lock file](https://carvel.dev/vendir/docs/v0.40.x/vendir-lock-spec/) which is generated by vendir and should be stored in source code.
+Renovate supports updating Helm Chart references and git references in vendir.yml via the [vendir](https://carvel.dev/vendir/) tool. Renovate requires the presence of a [vendir lock file](https://carvel.dev/vendir/docs/v0.40.x/vendir-lock-spec/) which is generated by vendir and should be stored in source code.
+
+## Helm Charts
```
Comment (viceice) [inline lib/modules/manager/vendir/readme.md:3]:
```suggestion
### Helm Charts
```

File: lib/modules/manager/vendir/readme.md [hunk 6]
```diff
@@ -32,3 +34,27 @@ directories:
 #### OCI
 
 Aliases for OCI registries are supported via the dockerfile/docker manager
+
+## Git
```
Comment (viceice) [inline lib/modules/manager/vendir/readme.md:38]:
```suggestion
### Git
```

File: lib/modules/manager/vendir/artifacts.spec.ts [hunk 7]
```diff
@@ -462,7 +462,6 @@ describe('modules/manager/vendir/artifacts', () => {
       fs.readLocalFile.mockResolvedValueOnce(vendirLockFile1);
       fs.getSiblingFileName.mockReturnValueOnce('vendir.lock.yml');
       fs.readLocalFile.mockResolvedValueOnce(vendirLockFile2);
-      fs.readLocalFile.mockResolvedValueOnce('0.35.0');
```
Comment (viceice) [inline lib/modules/manager/vendir/artifacts.spec.ts:465]: leftover from previous implementation? 😏

File: lib/modules/manager/vendir/extract.ts [hunk 8]
```diff
@@ -71,12 +91,22 @@ export function extractPackageFile(
...
-      deps.push({
-        ...dep,
-        depType: 'HelmChart',
-      });
+    if ('helmChart' in content && content.helmChart) {
+      const dep = extractHelmChart(content.helmChart, config.registryAliases);
+      if (dep) {
+        deps.push({
+          ...dep,
+          depType: 'HelmChart',
```
Comment (viceice) [inline lib/modules/manager/vendir/extract.ts:99]: should we move this to the extract functions? So we don't need to clone opjects here again? (perf)

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/manager/vendir/artifacts.spec.ts (hunk 7 above)
Comment (knechtionscoding) [reply lib/modules/manager/vendir/artifacts.spec.ts:465]: Yes. No longer needed.

File: lib/modules/manager/vendir/extract.ts (hunk 8 above)
Comment (knechtionscoding) [reply lib/modules/manager/vendir/extract.ts]: Should we? No idea. But I did if it speeds up performance. I don't think it will matter much, in this case, because for vendir the blocking thing will always be vendir sync, but happy to help where possible.

File: lib/modules/manager/vendir/extract.ts [hunk 9]
```diff
@@ -71,12 +94,20 @@ export function extractPackageFile(
...
-      deps.push({
-        ...dep,
-        depType: 'HelmChart',
-      });
+    if ('helmChart' in content && content.helmChart) {
+      const dep = extractHelmChart(content.helmChart, config.registryAliases);
+      if (dep) {
+        deps.push({
+          ...dep,
+        });
```
Comment (viceice) [inline lib/modules/manager/vendir/extract.ts:102]:
```suggestion
        deps.push(dep);
```

File: lib/modules/manager/vendir/extract.ts [hunk 10]
```diff
@@ -71,12 +94,20 @@ export function extractPackageFile(
...
+      }
+    } else if ('git' in content && content.git) {
+      const dep = extractGitSource(content.git);
+      if (dep) {
+        deps.push({
+          ...dep,
+        });
```
Comment (viceice) [inline lib/modules/manager/vendir/extract.ts:109]:
```suggestion
        deps.push(dep);
```

**Review by viceice — COMMENTED** [review]
😉

**Review by viceice — APPROVED** [review]
